src/components/protocol_handler/src/protocol_handler_impl.cc

```diff
@@ -1987,4 +2375,222 @@ uint8_t ProtocolHandlerImpl::SupportedSDLProtocolVersion() const {
+            "Adding AOA_USB to secondaryTransports for StartSessionAck");
+        secondaryTransports.push_back("AOA_USB");
+      }
+    } else if (transport_type.CompareIgnoreCase("Bluetooth")) {
+      if (device_is_ios) {
+        LOG4CXX_TRACE(
+            logger_,
+            "Adding IAP_BLUETOOTH to secondaryTransports for StartSessionAck");
+        secondaryTransports.push_back("IAP_BLUETOOTH");
+      }
+      if (device_is_android) {
+        LOG4CXX_TRACE(
+            logger_,
+            "Adding SPP_BLUETOOTH to secondaryTransports for StartSessionAck");
+        secondaryTransports.push_back("SPP_BLUETOOTH");
+      }
+    }
+    if (transport_type.CompareIgnoreCase("WiFi")) {
+      LOG4CXX_TRACE(
+          logger_,
+          "Adding TCP_WIFI to secondaryTransports for StartSessionAck");
+      secondaryTransports.push_back("TCP_WIFI");
+    }
+
+    it++;
+  }
+}
+
+void ProtocolHandlerImpl::generateServiceTransportsForStartSessionAck(
+    const std::vector<std::string>& service_transports,
+    const std::string& primary_connection_type,
+    const std::string& primary_transport_type,
+    const std::vector<std::string>& secondary_transport_types,
+    std::vector<int32_t>& serviceTransports) const {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  if (service_transports.size() == 0) {
+    LOG4CXX_TRACE(logger_,
+                  "Empty Service Transports. Allowing service to run on both "
+                  "connections");
+    serviceTransports.push_back(1);
+    serviceTransports.push_back(2);
+  } else {
+    bool fPrimaryAdded = false;
+    bool fSecondaryAdded = false;
+    std::vector<std::string>::const_iterator it = service_transports.begin();
+    for (; it != service_transports.end(); it++) {
+      const utils::custom_string::CustomString transport(*it);
+      LOG4CXX_TRACE(logger_,
+                    "Service Allowed to run on " << transport.c_str()
+                                                 << " transport");
+
+      if (!fPrimaryAdded &&
+          (transport.CompareIgnoreCase(primary_connection_type.c_str()) ||
+           (transport.CompareIgnoreCase("IAP_USB") &&
+            primary_transport_type.compare("USB") == 0))) {
+        LOG4CXX_TRACE(logger_, "Service allowed on primary transport");
+        serviceTransports.push_back(1);
+        fPrimaryAdded = true;
+      }
+
+      if (!fSecondaryAdded) {
+        const utils::custom_string::CustomString transport_type(
+            transportTypeFromTransport(transport));
+        std::vector<std::string>::const_iterator found =
+            std::find_if(secondary_transport_types.begin(),
+                         secondary_transport_types.end(),
+                         [&](const std::string& secondary_transport_type) {
+                           return transport_type.CompareIgnoreCase(
+                               secondary_transport_type.c_str());
+                         });
+        if (found != secondary_transport_types.end()) {
```

Comment:
Sort of a fundamental question to this feature, but what if there are several secondary transports are available for a given transport, but only one of them can support video/audio in the config? The current logic would imply that all of the secondary transports support video/audio if `2` is present in this array, but that is not the case. There is no way to determine which secondary transports support these services from the app's perspective if multiple are available.

Reply:
Good question. Actually this is something we missed during the discussion of protocol design - my initial idea was to list up all of transport types in `audioServiceTransports` and `videoServiceTransports`, but later the params are simplified just to contain "1" and/or "2". So as you pointed out, it is possible that `secondaryTransports` has more than one transport candidates, and only some of those are allowed for audio/video streamings.

If we need to resolve the issue, we need to update SDL protocol again, which I think will require another proposal.

Thinking of the initial motivation however, I don't think we have to support that case in high priority. The secondary transport is dedicated for audio/video streamings, so there is no point to include a low-bandwidth transport (Bluetooth in my mind) in the `secondaryTransports` parameter.